File: android/src/main/java/com/github/musicode/photopicker/RNTPhotoPickerModule.java
package com.github.musicode.photopicker;

import android.app.Activity;
import android.content.pm.PackageManager;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.widget.ImageView;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.github.herokotlin.photopicker.PhotoPickerActivity;
import com.github.herokotlin.photopicker.PhotoPickerCallback;
import com.github.herokotlin.photopicker.PhotoPickerConfiguration;
import com.github.herokotlin.photopicker.model.PickedAsset;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function5;

public class RNTPhotoPickerModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;

    public RNTPhotoPickerModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    private static Function5<ImageView, String, Integer, Integer, Function1, Unit> loader;

    public static void setImageLoader(Function5<ImageView, String, Integer, Integer, Function1, Unit> loader) {
        RNTPhotoPickerModule.loader = loader;
    }

    @Override
    public String getName() {
        return "RNTPhotoPicker";
    }

    @ReactMethod
    public void open(ReadableMap options, final Promise promise) {

        PhotoPickerConfiguration configuration = new PhotoPickerConfiguration() {
            @Override
            public void loadAsset(ImageView imageView, String url, int loading, int error, Function1<? super Boolean, Unit> onComplete) {
                loader.invoke(imageView, url, loading, error, onComplete);
            }

            @Override
            public boolean requestPermissions(Activity activity, List<String> permissions, int requestCode) {
                List<String> list = new ArrayList<>();

                for (String permission: permissions) {
                    if (ContextCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
                        list.add(permission);
                    }
                }

                if (list.size() > 0) {
                    ActivityCompat.requestPermissions(activity, list.toArray(new String[list.size()]), requestCode);
                    return false;
                }

                return true;
            }
        };

        configuration.setCountable(options.getBoolean("countable"));
        configuration.setMaxSelectCount(options.getInt("maxSelectCount"));
        configuration.setRawButtonVisible(options.getBoolean("rawButtonVisible"));

        if (options.hasKey("imageMinWidth") && options.getInt("imageMinWidth") > 0) {
            configuration.setImageMinWidth(options.getInt("imageMinWidth"));
        }
        if (options.hasKey("imageMinHeight") && options.getInt("imageMinHeight") > 0) {
            configuration.setImageMinHeight(options.getInt("imageMinHeight"));
        }
        if (options.hasKey("cancelButtonTitle")) {
            configuration.setCancelButtonTitle(options.getString("cancelButtonTitle"));
        }
        if (options.hasKey("rawButtonTitle")) {
            configuration.setRawButtonTitle(options.getString("rawButtonTitle"));
        }
        if (options.hasKey("submitButtonTitle")) {
            configuration.setSubmitButtonTitle(options.getString("submitButtonTitle"));
        }

        PhotoPickerCallback callback = new PhotoPickerCallback() {

            @Override
            public void onCancel(@NotNull Activity activity) {
                activity.finish();
                promise.reject("-1", "cancel");
            }

            @Override
            public void onPermissionsNotGranted(@NotNull Activity activity) {
                activity.finish();
                promise.reject("1", "has no permissions");
            }

            @Override
            public void onPermissionsDenied(@NotNull Activity activity) {
                activity.finish();
                promise.reject("2", "you denied the requested permissions.");
            }

            @Override
            public void onExternalStorageNotWritable(@NotNull Activity activity) {
                activity.finish();
                promise.reject("3", "external storage is not writable");
            }

            @Override
            public void onPermissionsGranted(@NotNull Activity activity) {

            }

            @Override
            public void onSubmit(@NotNull Activity activity, List<PickedAsset> list) {

                activity.finish();

                WritableArray array = Arguments.createArray();

                for (int i = 0; i < list.size(); i++) {
                    WritableMap map = Arguments.createMap();
                    PickedAsset asset = list.get(i);

                    map.putString("path", asset.getPath());
                    map.putInt("size", asset.getSize());
                    map.putInt("width", asset.getWidth());
                    map.putInt("height", asset.getHeight());
                    map.putBoolean("isVideo", asset.isVideo());
                    map.putBoolean("isRaw", asset.isRaw());

                    array.pushMap(map);
                }

                promise.resolve(array);

            }
        };

        PhotoPickerActivity.Companion.setConfiguration(configuration);
        PhotoPickerActivity.Companion.setCallback(callback);

        PhotoPickerActivity.Companion.newInstance(reactContext.getCurrentActivity());

    }

}

File: android/src/main/java/com/github/musicode/photopicker/RNTPhotoPickerModule.kt
package com.github.musicode.photopicker

import android.app.Activity
import android.widget.ImageView

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.github.herokotlin.photopicker.PhotoPickerActivity
import com.github.herokotlin.photopicker.PhotoPickerCallback
import com.github.herokotlin.photopicker.PhotoPickerConfiguration
import com.github.herokotlin.photopicker.model.PickedAsset

class RNTPhotoPickerModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    companion object {

        private lateinit var loader: (imageView: ImageView, url: String, loading: Int, error: Int, onComplete: (Boolean) -> Unit) -> Unit

        fun init(loader: (imageView: ImageView, url: String, loading: Int, error: Int, onComplete: (Boolean) -> Unit) -> Unit) {
            this.loader = loader
        }

    }

    override fun getName(): String {
        return "RNTPhotoPicker"
    }

    @ReactMethod
    fun open(options: ReadableMap, promise: Promise) {

        val configuration = object : PhotoPickerConfiguration() {
            override fun loadAsset(imageView: ImageView, url: String, loading: Int, error: Int, onComplete: (Boolean) -> Unit) {
                loader.invoke(imageView, url, loading, error, onComplete)
            }
        }

        configuration.countable = options.getBoolean("countable")
        configuration.maxSelectCount = options.getInt("maxSelectCount")
        configuration.showOriginalButton = options.getBoolean("showOriginalButton")

        if (options.hasKey("imageMinWidth") && options.getInt("imageMinWidth") > 0) {
            configuration.imageMinWidth = options.getInt("imageMinWidth")
        }
        if (options.hasKey("imageMinHeight") && options.getInt("imageMinHeight") > 0) {
            configuration.imageMinHeight = options.getInt("imageMinHeight")
        }
        if (options.hasKey("cancelButtonTitle")) {
            configuration.cancelButtonTitle = options.getString("cancelButtonTitle") as String
        }
        if (options.hasKey("originalButtonTitle")) {
            configuration.originalButtonTitle = options.getString("originalButtonTitle") as String
        }
        if (options.hasKey("submitButtonTitle")) {
            configuration.submitButtonTitle = options.getString("submitButtonTitle") as String
        }

        val callback = object : PhotoPickerCallback {

            override fun onCancel(activity: Activity) {
                activity.finish()
                promise.reject("-1", "cancel")
            }

            override fun onSubmit(activity: Activity, assetList: List<PickedAsset>) {

                activity.finish()

                val array = Arguments.createArray()

                for (i in assetList.indices) {
                    val map = Arguments.createMap()
                    val (path, _, base64, width, height, size, isVideo, isOriginal) = assetList[i]

                    map.putString("path", path)
                    map.putString("base64", base64)
                    map.putInt("size", size)
                    map.putInt("width", width)
                    map.putInt("height", height)
                    map.putBoolean("isVideo", isVideo)
                    map.putBoolean("isOriginal", isOriginal)

                    array.pushMap(map)
                }

                promise.resolve(array)

            }
        }

        PhotoPickerActivity.configuration = configuration
        PhotoPickerActivity.callback = callback

        PhotoPickerActivity.newInstance(reactContext.currentActivity!!)

    }

}
